### DISCUSSION TITLE
the sql is too slow

### DISCUSSION THREAD

**Discussion by zenghao306** [filecoin-project/lily#334] [Q&A] [answered]
When there has many data,in the func [FindGasOutputsMessages] the sql just as this :
	_, err := tx.QueryContext(ctx, &list, `
		SELECT pm.height, pm.cid, m.from, m.to, m.size_bytes, m.nonce, m.value,
			   m.gas_fee_cap, m.gas_premium, m.gas_limit, m.method,
			   r.state_root, r.exit_code,r.gas_used, bh.parent_base_fee
		FROM visor_processing_messages pm
		JOIN receipts r ON pm.cid = r.message -- don't join receipts on height since it's the height of the receipt
		JOIN messages m ON pm.cid = m.cid AND pm.height = m.height
		JOIN block_messages bm on pm.cid = bm.message AND pm.height = bm.height
		JOIN block_headers bh on bm.block = bh.cid AND bm.height = bh.height
		WHERE pm.gas_outputs_completed_at IS null AND
		      pm.height >= ? AND pm.height <= ?
		ORDER BY pm.height DESC
		LIMIT ?
`, minHeight, maxHeight, batchSize)

this SQL is too slow，Can you have a way to increase the speed？

**Comment by frrist** [comment, marked as answer]
We recently add a new command called "walk" which should work better for you.

You can run it like this to index message gas output states:

```
visor walk --from 1000 --to 2000 --tasks=messages 
```

You can include multiple tasks by comma separating them. The full list is here https://github.com/filecoin-project/sentinel-visor/blob/master/chain/indexer.go#L23
